In [3]:
import pandas as pd
import os

# Reading the values and display

In [4]:
df = pd.read_csv(r"dataset.csv")
df

Roll CT-1 (Marks: 20) CT-2 (Marks: 20) CT-3 (Marks: 20) CT-4 (Marks: 20)  \
0    2.0             12.5               15               14               18   
1    3.0                9               10               17               11   
2    4.0               10             15.5             12.5              6.5   
3    5.0               14               14               18                0   
4    6.0             14.5             15.5               15             15.5   
..   ...              ...              ...              ...              ...   
58  60.0               12               20               14               18   
59  61.0             19.5             17.5               17                0   
60  62.0                0                4               13             12.5   
61  63.0                0                4               16              7.5   
62  64.0               18             16.5                0               15   

   Class Test (Total of best 3, Marks: 60) Attendance (Marks: 15)  \
0                                       47                     15   
1                                       38                     15   
2                                       38                     15   
3                                       46                     15   
4                                       46                     15   
..                                     ...                    ...   
58                                      52                     15   
59                                      54                     15   
60                                    29.5                     15   
61                                    27.5                   13.5   
62                                    49.5                   13.5   

   Observation (Marks: 15) Total (Marks: 90)  Exam Mark (105)  \
0                     13.5              75.5             45.0   
1                     13.5              66.5             28.0   
2                     13.5              66.5             26.5   
3                       15                76             46.0   
4                       15                76             26.0   
..                     ...               ...              ...   
58                    13.5              80.5             51.0   
59                      15                84             32.0   
60                    12.5                57             50.0   
61                    12.5              53.5             16.0   
62                    13.5              76.5             12.0   

    Exam Mark (105).1  Mark total 300  
0                46.5           167.0  
1                44.0           138.5  
2                43.0           136.0  
3                44.0           166.0  
4                74.5           176.5  
..                ...             ...  
58               66.0           197.5  
59               84.0           200.0  
60               77.5           184.5  
61               39.0           108.5  
62               45.0           133.5  

[63 rows x 12 columns]

In [5]:
df.fillna(0,inplace=True)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Data columns (total 12 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Roll                                     63 non-null     float64
 1   CT-1 (Marks: 20)                         63 non-null     object 
 2   CT-2 (Marks: 20)                         63 non-null     object 
 3   CT-3 (Marks: 20)                         63 non-null     object 
 4   CT-4 (Marks: 20)                         63 non-null     object 
 5   Class Test (Total of best 3, Marks: 60)  63 non-null     object 
 6   Attendance (Marks: 15)                   63 non-null     object 
 7   Observation (Marks: 15)                  63 non-null     object 
 8   Total (Marks: 90)                        63 non-null     object 
 9   Exam Mark (105)                          63 non-null     float64
 10  Exam Mark (105).1                        63 non-null

# selecting the required columns

In [7]:
df['Attendance (Marks: 15)'].value_counts()

15                        49
0                          6
13.5                       4
Attendance (Marks: 15)     3
10.5                       1
Name: Attendance (Marks: 15), dtype: int64

# grouping and splitting based on attendance




In [8]:
g=df.groupby('Attendance (Marks: 15)')

In [9]:
df['Mark total 300'].mean()

141.70634920634922

In [10]:
g['Mark total 300'].mean()

Attendance (Marks: 15)
0                           0.000000
10.5                      136.500000
13.5                      148.875000
15                        167.255102
Attendance (Marks: 15)      0.000000
Name: Mark total 300, dtype: float64

In [11]:
g['Mark total 300'].median()

Attendance (Marks: 15)
0                           0.00
10.5                      136.50
13.5                      154.25
15                        166.00
Attendance (Marks: 15)      0.00
Name: Mark total 300, dtype: float64

In [12]:
df['Mark total 300'].median()

160.0

# with reference to the total marks scored students with both 13.5 and 15 marks in attendance scored above the average but only students with 15 marks in attendance crossed the median

In [13]:
df['CT-1 (Marks: 20)']=pd.to_numeric(df['CT-1 (Marks: 20)'],errors='coerce')
df['CT-2 (Marks: 20)']=pd.to_numeric(df['CT-2 (Marks: 20)'],errors='coerce')
df['CT-3 (Marks: 20)']=pd.to_numeric(df['CT-3 (Marks: 20)'],errors='coerce')
df['CT-4 (Marks: 20)']=pd.to_numeric(df['CT-4 (Marks: 20)'],errors='coerce')
df['Class Test (Total of best 3, Marks: 60)']=pd.to_numeric(df['Class Test (Total of best 3, Marks: 60)'],errors='coerce')

In [14]:
df['CT-1 (Marks: 20)'].mean()

9.866666666666667

In [15]:
df['CT-1 (Marks: 20)'].median()

11.25

In [16]:
g['CT-1 (Marks: 20)'].mean()

Attendance (Marks: 15)
0                          0.000000
10.5                      16.000000
13.5                       5.750000
15                        11.285714
Attendance (Marks: 15)          NaN
Name: CT-1 (Marks: 20), dtype: float64

In [17]:
g['CT-1 (Marks: 20)'].median()

Attendance (Marks: 15)
0                          0.0
10.5                      16.0
13.5                       2.5
15                        12.0
Attendance (Marks: 15)     NaN
Name: CT-1 (Marks: 20), dtype: float64

# class test 1 results seem to have no relation with attendance with students with the least attendance scoring the highest 

In [18]:
df['CT-2 (Marks: 20)'].mean()

13.5

In [19]:
df['CT-2 (Marks: 20)'].median()

15.0

In [20]:
g['CT-2 (Marks: 20)'].mean()

Attendance (Marks: 15)
0                          0.00000
10.5                      15.50000
13.5                      12.12500
15                        15.22449
Attendance (Marks: 15)         NaN
Name: CT-2 (Marks: 20), dtype: float64

In [21]:
g['CT-2 (Marks: 20)'].median()

Attendance (Marks: 15)
0                          0.0
10.5                      15.5
13.5                      14.0
15                        15.0
Attendance (Marks: 15)     NaN
Name: CT-2 (Marks: 20), dtype: float64

# once again it seems that students with the least attendance seem to be doing better at class tests and students with mediocre attendance performed poorly again

In [22]:
df['CT-3 (Marks: 20)'].mean()

12.583333333333334

In [23]:
df['CT-3 (Marks: 20)'].median()

14.0

In [24]:
g['CT-3 (Marks: 20)'].mean()

Attendance (Marks: 15)
0                          0.000000
10.5                      15.000000
13.5                      10.500000
15                        14.244898
Attendance (Marks: 15)          NaN
Name: CT-3 (Marks: 20), dtype: float64

In [25]:
g['CT-3 (Marks: 20)'].median()

Attendance (Marks: 15)
0                          0.0
10.5                      15.0
13.5                      13.0
15                        14.5
Attendance (Marks: 15)     NaN
Name: CT-3 (Marks: 20), dtype: float64

# similar trend as the last  two class tests 

In [26]:
df['CT-4 (Marks: 20)'].mean()

8.875

In [27]:
df['CT-4 (Marks: 20)'].median()

8.75

In [28]:
g['CT-4 (Marks: 20)'].mean()

Attendance (Marks: 15)
0                          0.000000
10.5                       0.000000
13.5                      11.625000
15                         9.918367
Attendance (Marks: 15)          NaN
Name: CT-4 (Marks: 20), dtype: float64

In [29]:
g['CT-4 (Marks: 20)'].median()

Attendance (Marks: 15)
0                          0.0
10.5                       0.0
13.5                      12.0
15                        10.5
Attendance (Marks: 15)     NaN
Name: CT-4 (Marks: 20), dtype: float64

# its clearly visible that quite a few students did not attend this test since only the best of 3  was considered for the internals

In [30]:
df['Class Test (Total of best 3, Marks: 60)'].mean()

38.93333333333333

In [31]:
df['Class Test (Total of best 3, Marks: 60)'].median()

43.5

In [32]:
g['Class Test (Total of best 3, Marks: 60)'].mean()

Attendance (Marks: 15)
0                          0.000000
10.5                      46.500000
13.5                      38.750000
15                        43.561224
Attendance (Marks: 15)          NaN
Name: Class Test (Total of best 3, Marks: 60), dtype: float64

In [43]:
g['Class Test (Total of best 3, Marks: 60)'].median()

Attendance (Marks: 15)
0                          0.0
10.5                      46.5
13.5                      39.0
15                        44.0
Attendance (Marks: 15)     NaN
Name: Class Test (Total of best 3, Marks: 60), dtype: float64

# overall the students with least attendance have come out on top for the class test marks 

In [34]:
df['Exam Mark (105)'].mean()

30.833333333333332

In [35]:
df['Exam Mark (105)'].median()

30.0

In [36]:
g['Exam Mark (105)'].mean()

Attendance (Marks: 15)
0                          0.000000
10.5                       9.000000
13.5                      26.500000
15                        37.295918
Attendance (Marks: 15)     0.000000
Name: Exam Mark (105), dtype: float64

In [37]:
g['Exam Mark (105)'].median()

Attendance (Marks: 15)
0                          0.0
10.5                       9.0
13.5                      26.5
15                        38.0
Attendance (Marks: 15)     0.0
Name: Exam Mark (105), dtype: float64

# in the final exam marks however there is a clear distinction in the performance based on atendance

In [38]:
df['Exam Mark (105).1'].mean()

49.317460317460316

In [40]:
df['Exam Mark (105).1'].median()

55.0

In [41]:
g['Exam Mark (105).1'].mean()

Attendance (Marks: 15)
0                          0.000000
10.5                      57.000000
13.5                      56.875000
15                        57.602041
Attendance (Marks: 15)     0.000000
Name: Exam Mark (105).1, dtype: float64

In [42]:
g['Exam Mark (105).1'].median()

Attendance (Marks: 15)
0                          0.0
10.5                      57.0
13.5                      56.0
15                        60.0
Attendance (Marks: 15)     0.0
Name: Exam Mark (105).1, dtype: float64

# this distinction in performace is reflective of the attendance in class where students with the highest attendance perfomed better

# also something thats clearly visible in the fianl result